extension/extensionlimiter/README.md

```diff
@@ -0,0 +1,305 @@
+configured through middleware and/or directly by pipeline components.
+
+## Overview
+
+This package defines two primary limiter **kinds**, which have
+different interfaces:
+
+- **Rate Limiters**: Control time-based limits on quantities such as
+  bytes or items per second.
+- **Resource Limiters**: Manage physical limits on quantities such as
+  concurrent requests or memory usage.
+
+Both limiter kinds are unified through the `LimiterWrapper` interface,
+which simplifies consumers in most cases by providing a consistent
+`LimitCall` interface.
+
+A limiter is **saturated** by definition when a limit is completely
+overloaded, generally it means a limit request of any size would fail
+at that moment and should be taken as a strong signal to stop
+accepting requests.
+
+Each kind of limiter as well as the wrapper type have corresponding
+**provider** interfaces that return a limiter instance based on a
+weight key or keys.
+
+Weight keys describe the standard limiting dimensions. There are
+currently four standard weight keys: network bytes, request count,
+request items, and memory size.
+
+## Key Interfaces
+
+- `LimiterWrapper`: Provides a callback-based limiting interface that
+  works with both rate and resource limiters, has a `LimitCall` method,
+  plus a provider type.
+- `RateLimiter`: Applies time-based limits, has a `Limit` method,
+  plus provider type.
+- `ResourceLimiter`: Manages physical resource limits, has
+  an `Acquire` method and a corresponding `ReleaseFunc`,
+  plus a provider type.
+- `Limiter`: Any of the above, has a `MustDeny` method.
+
+### Limiter helpers
+
+The `limiterhelper` subpackage provides:
+
+- Consumer wrappers apply limits to a collector pipeline (e.g.,
+  `NewLimitedLogs` for a limiter combined with `consumer.NewLogs`)
+- Multi-limiter combinators: for simple combined limiter functionality
+- Middleware conversion utilities: convert middleware configurations
+  to limiter providers.
+
+## Recommendations
+
+For general use cases, prefer the `LimiterWrapper` interface with its
+callback-based approach because it is agnostic to the difference between
+rate and resource limiters.
+
+Use the direct `RateLimiter` or `ResourceLimiter` interfaces only in
+special cases where control flow can't be easily scoped.
+
+Middleware configuration typically automates the configuration of
+network bytes and request count weight keys relatively early in a
+pipeline.  Receivers are responsible for limiting request items and
+memory size through one of the available helpers.
+
+Processors can apply limiters for specific reasons, for example to
+apply limits in data-dependent ways.  Exporters can apply limiters for
+the same reasons, for example to apply limits in destination-dependent
+ways.
+
+### Limiter blocking and failing
+
+Limiters implementations MAY block the request or fail immediately,
+subject to internal logic. A limiter aims to avoid waste, which
```

Comment:
What is the internal logic? Are you saying this is a property of the limiter, rather than the caller?

If we made it caller-defined, then I think we could get rid of the MustDeny call: instead you would make a non-blocking request for 0 items. If the limiter is saturated, that would return an error; otherwise it would return success without affecting the capacity.

Reply:
I have raised a question about non-blocking methods in the open-questions section of the README.